File: code/app-api/app/rest_controllers/items.py
from ..rest_base import RestControllerBase, RestEndpoint
from ..auth import is_authenticated
from .. import couchbase as cb, env
import uuid
from fastapi import HTTPException
from fastapi.responses import Response
from pydantic import BaseModel

class ParamsItemCreate(BaseModel):
    name: str

class ParamsItemResponse(BaseModel):
    id: str
    name: str

class ItemsController(RestControllerBase):
    def __init__(self, context):
        super().__init__(context)

    @RestEndpoint("GET", "", auth=is_authenticated)
    async def index(self) -> list[ParamsItemResponse]:
        result = cb.exec(
            env.get_couchbase_conf(),
            f"SELECT name, META().id FROM {env.get_couchbase_bucket()}._default.items"
        )
        return [ParamsItemResponse(id=r['id'], name=r['name']) for r in result]

    @RestEndpoint("POST", "", auth=is_authenticated)
    async def create(self, params: ParamsItemCreate) -> ParamsItemResponse:
        id = str(uuid.uuid1())
        cb.insert(env.get_couchbase_conf(),
                  cb.DocSpec(bucket=env.get_couchbase_bucket(),
                             collection='items',
                             key=id,
                             data={'name': params.name}))
        return ParamsItemResponse(id=id, name=params.name)

    @RestEndpoint("DELETE", "/{id}", auth=is_authenticated)
    async def remove(self, id: str):
        try:
            cb.remove(env.get_couchbase_conf(),
                      cb.DocRef(bucket=env.get_couchbase_bucket(),
                                collection='items',
                                key=id))
            return Response(status_code=204)
        except Exception as e:
            raise HTTPException(status_code=404, detail=f"Item with id {id} not found")

File: code/web-app/src/rest/modules/items.ts
import { ApiClientInterface } from '../types';

export interface Item {
    id: string;
    data: ItemData;
}

export interface ItemData {
    name: string;
}

export function create(client: ApiClientInterface) {
    return {
        list(): Promise<Item[]> {
            const on_error = (messages: string[]) => {
                console.error('Error in items list API:', messages);
            };
            return client.get<Item[]>('/api/items', on_error);
        },

        create(item: Omit<Item, 'id'>): Promise<Item> {
            const on_error = (messages: string[]) => {
                console.error('Error in items create API:', messages);
            };
            return client.post<Item>('/api/items', on_error, item);
        },

        remove(id: string): Promise<null> {
            const on_error = (messages: string[]) => {
                console.error('Error in items remove API:', messages);
            };
            return client.delete(`/api/items/${id}`, on_error);
        }
    };
}
